Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/Update.ts

```diff
@@ -3,96 +3,273 @@
+  const { acknowledged, matchedCount } = await mongoCollection.replaceOne({ _id: meadowlarkId, documentUuid }, document, {
+    session,
+  });
+
+  // Check for general MongoDB problems
+  if (!acknowledged) {
+    const msg =
+      'mongoCollection.replaceOne returned acknowledged: false, indicating a problem with write concern configuration';
+    Logger.error(`${moduleName}.tryUpdateByReplacement`, traceId, msg);
+    return { response: 'UNKNOWN_FAILURE' };
+  }
+
+  // If there was a match on replace, success
+  if (matchedCount > 0) return { response: 'UPDATE_SUCCESS' };
+
+  return null;
+}
+
+/**
+ * Attempts to update the document, by replacement if there is no identity (meadowlarkId) change, or by
+ * by deleting and reinserting if there is an identity change. Delete + reinsert allows for identity change because
+ * identity (meadowlarkId) is the MongoDB _id.
+ *
+ * This function expects Session to have an active transaction. Aborting the transaction on error is left to the caller.
+ */
+async function updateAllowingIdentityChange(
   document: MeadowlarkDocument,
+  updateRequest: UpdateRequest,
+  mongoCollection: Collection<MeadowlarkDocument>,
   session: ClientSession,
 ): Promise<UpdateResult> {
-  let updateResult: UpdateResult = { response: 'UNKNOWN_FAILURE' };
-  // Perform the document update
-  Logger.debug(`${moduleName}.updateDocumentById: Updating document uuid ${documentUuid}`, traceId);
-  const documentToUpdate = await mongoCollection.findOne({ documentUuid }, { projection: { _id: 1 } });
-  // eslint-disable-next-line no-underscore-dangle
-  if (documentToUpdate !== null && documentToUpdate._id !== meadowlarkId) {
-    // This function cannot update a natural key.
-    updateResult = {
-      response: 'UPDATE_FAILURE_NATURAL_KEY',
-    };
-  } else {
-    const { acknowledged, matchedCount } = await mongoCollection.replaceOne({ _id: meadowlarkId, documentUuid }, document, {
-      session,
-    });
-    if (acknowledged) {
-      updateResult = {
-        response: matchedCount > 0 ? 'UPDATE_SUCCESS' : 'UPDATE_FAILURE_NOT_EXISTS',
-      };
-    } else {
-      const msg =
-        'mongoCollection.replaceOne returned acknowledged: false, indicating a problem with write concern configuration';
-      Logger.error(`${moduleName}.updateDocumentById`, traceId, msg);
-    }
+  const { documentUuid, resourceInfo, traceId, security } = updateRequest;
+
+  // Optimize by trying a replacement update, which will succeed if there is no identity change
+  const tryUpdateByReplacementResult: UpdateResult | null = await tryUpdateByReplacement(
+    document,
+    updateRequest,
+    mongoCollection,
+    session,
+  );
+
+  if (tryUpdateByReplacementResult != null) {
+    // Ensure referenced documents are not modified in other transactions
+    await writeLockReferencedDocuments(mongoCollection, document.outboundRefs, session);
+    return tryUpdateByReplacementResult;
```

Comment:
Only write-lock if the attempt succeeds. Otherwise, let upsert do the write locking.